File: filly-core/src/widgets/tree.rs
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, ListItem, Rect, RenderNode, RenderTree, TextStyle, TreeNode,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct TreeWidget {
    pub title: String,
    pub nodes: Vec<TreeNode>,      // original tree (expanded state lives here)
    pub cursor: usize,             // index into the flat visible list
    dirty: bool,
}

impl TreeWidget {
    pub fn new(title: String, nodes: Vec<TreeNode>) -> Self {
        Self { title, nodes, cursor: 0, dirty: true }
    }

    fn flatten(&self) -> Vec<(usize, String, bool)> {
        let mut out = Vec::new();
        Self::flatten_rec(&self.nodes, 0, &mut out);
        out
    }

    fn flatten_rec(nodes: &[TreeNode], depth: usize, out: &mut Vec<(usize, String, bool)>) {
        for node in nodes {
            let marker = if node.children.is_empty() {
                "  "
            } else if node.expanded {
                "▼ "
            } else {
                "▶ "
            };
            out.push((depth, format!("{}{}", marker, node.label), node.children.is_empty()));
            if node.expanded {
                Self::flatten_rec(&node.children, depth + 1, out);
            }
        }
    }

    fn toggle_expand(&mut self) {
        let flat = self.flatten();
        if let Some((_, _, is_leaf)) = flat.get(self.cursor) {
            if *is_leaf { return; }
            let mut remaining = self.cursor;
            if Self::toggle_at(&mut self.nodes, &mut remaining) {
                self.dirty = true;
            }
        }
    }

    fn toggle_at(nodes: &mut [TreeNode], idx: &mut usize) -> bool {
        for node in nodes.iter_mut() {
            if *idx == 0 {
                if !node.children.is_empty() {
                    node.expanded = !node.expanded;
                    return true;
                }
                return false;
            }
            *idx -= 1;
            if node.expanded {
                if Self::toggle_at(&mut node.children, idx) {
                    return true;
                }
            }
        }
        false
    }
}

impl Widget for TreeWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = ((area.width as f32 * 0.7) as u16).min(area.width.saturating_sub(2));
        let box_h = ((area.height as f32 * 0.8) as u16).min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;
        let flat = self.flatten();
        let items: Vec<ListItem> = flat.iter().map(|(depth, label, _)| {
            ListItem {
                label: format!("{:indent$}{}", "", label, indent = depth * 2),
                meta: None,
            }
        }).collect();

        let mut children = Vec::new();
        children.push(RenderNode::Text {
            rect: Rect::new(1, 0, box_w - 2, 1),
            content: self.title.clone(),
            alignment: Alignment::Center,
            role: None, label: None, description: None, style: TextStyle { bold: true, fg: Some("green".into()), ..TextStyle::normal() },
        });
        children.push(RenderNode::List {
            rect: Rect::new(1, 1, box_w - 2, box_h - 3),
            items,
            selected: Some(self.cursor.min(flat.len().saturating_sub(1))),
            highlight_style: TextStyle::selected(), role: None, label: None, description: None,
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, box_h - 2, box_w - 2, 1),
            content: "Up/Down:move  Enter/Space:expand  Esc:cancel".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(), role: None, label: None, description: None,
            children,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        match event {
            Event::Key(code, _) => match code {
                KeyCode::Esc => EventResult::Response(WidgetResponse {
                    result: None, cancelled: true, error: None,
                }),
                KeyCode::Up => {
                    if self.cursor > 0 { self.cursor -= 1; self.dirty = true; }
                    EventResult::Handled
                }
                KeyCode::Down => {
                    let len = self.flatten().len();
                    if self.cursor + 1 < len { self.cursor += 1; self.dirty = true; }
                    EventResult::Handled
                }
                KeyCode::Enter | KeyCode::Char(' ') => {
                    self.toggle_expand();
                    EventResult::Handled
                }
                _ => EventResult::Unhandled,
            },
            _ => EventResult::Unhandled,
        }
    }

    fn is_dirty(&self) -> bool { self.dirty }
    fn clear_dirty(&mut self) { self.dirty = false; }
}

File: filly-core/widgets/tree.c
#include <stdio.h>
#include "tree.h"
#include <stdlib.h>
#include <string.h>

typedef struct {
    char *title;
    TreeNode *nodes;
    int node_count;
    int *selected_path;
    int path_len;
    int cursor;
    int *flat_indices;
    int flat_count;
    bool dirty;
} TreeData;

static void flatten(TreeNode **nodes, int count, int *path, int *out_indices, int *out_count, int *cursor_ptr) {
    for (int i = 0; i < count; i++) {
        out_indices[(*out_count)++] = *cursor_ptr;
        (*cursor_ptr)++;
        if (nodes[i]->expanded && nodes[i]->child_count > 0)
            flatten(nodes[i]->children, nodes[i]->child_count, path, out_indices, out_count, cursor_ptr);
    }
}

static void tree_render(Widget *self, Rect area, RenderTree *out) {
    TreeData *d = (TreeData *)(self + 1);
    int total_nodes = 0;
    for (int i = 0; i < d->node_count; i++) total_nodes += d->nodes[i].child_count + 1;
    d->flat_indices = realloc(d->flat_indices, total_nodes * sizeof(int));
    d->flat_count = 0;
    int cursor = 0;
    for (int i = 0; i < d->node_count; i++) {
        d->flat_indices[d->flat_count++] = i;
        if (d->nodes[i].expanded && d->nodes[i].child_count > 0) {
            for (int j = 0; j < d->nodes[i].child_count; j++) {
                d->flat_indices[d->flat_count++] = d->node_count + j;
            }
        }
    }

    memset(out, 0, sizeof(*out));
    int box_w = (int)(area.w * 0.7f);
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_h = (int)(area.h * 0.8f);
    if (box_h > area.h - 2) box_h = area.h - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    RenderTree *children = calloc(3 + 1, sizeof(RenderTree));
    children[0].type = RNODE_TEXT;
    children[0].rect = rect_new(1, 0, box_w - 2, 1);
    children[0].text.content = strdup(d->title);
    children[0].text.align = ALIGN_CENTER;
    children[0].text.style = textstyle_selected();

    children[1].type = RNODE_LIST;
    children[1].rect = rect_new(1, 1, box_w - 2, box_h - 3);
    children[1].list.item_count = d->flat_count;
    children[1].list.items = malloc(d->flat_count * sizeof(ListItem));
    for (int i = 0; i < d->flat_count; i++) {
        int idx = d->flat_indices[i];
        const char *label = idx < d->node_count ? d->nodes[idx].label : "?";
        children[1].list.items[i] = listitem_new(label);
    }
    children[1].list.selected = d->cursor;
    children[1].list.highlight = textstyle_selected();

    children[2].type = RNODE_TEXT;
    children[2].rect = rect_new(1, box_h - 2, box_w - 2, 1);
    children[2].text.content = strdup("Up/Down:move  Enter/Space:expand  Esc:cancel");
    children[2].text.align = ALIGN_CENTER;
    children[2].text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = 3;
}

static EventResult tree_handle_event(Widget *self, Event *ev, Backend *backend) {
    TreeData *d = (TreeData *)(self + 1);
    if (ev->type != EVENT_KEY) return event_result_unhandled();
    switch (ev->code) {
        case KEY_ESC:
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        case KEY_UP: if (d->cursor > 0) d->cursor--; d->dirty = true; return event_result_handled();
        case KEY_DOWN: if (d->cursor + 1 < d->flat_count) d->cursor++; d->dirty = true; return event_result_handled();
        case KEY_ENTER:
        case KEY_CHAR:
            if (ev->code == KEY_CHAR && ev->ch != ' ') return event_result_unhandled();
            if (d->cursor < d->flat_count) {
                int idx = d->flat_indices[d->cursor];
                if (idx < d->node_count && d->nodes[idx].child_count > 0) {
                    d->nodes[idx].expanded = !d->nodes[idx].expanded;
                    d->dirty = true;
                }
            }
            return event_result_handled();
        default: return event_result_unhandled();
    }
}

static bool tree_is_dirty(Widget *self) { return ((TreeData *)(self + 1))->dirty; }
static void tree_clear_dirty(Widget *self) { ((TreeData *)(self + 1))->dirty = false; }
static void tree_destroy(Widget *self) {
    TreeData *d = (TreeData *)(self + 1);
    free(d->title); free(d->flat_indices);
}

Widget *tree_widget_new(const char *title, TreeNode *nodes, int node_count) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(TreeData));
    w->vtable.render = tree_render;
    w->vtable.handle_event = tree_handle_event;
    w->vtable.is_dirty = tree_is_dirty;
    w->vtable.clear_dirty = tree_clear_dirty;
    w->vtable.destroy = tree_destroy;
    TreeData *d = (TreeData *)(w + 1);
    d->title = strdup(title);
    d->nodes = nodes;
    d->node_count = node_count;
    d->cursor = 0;
    d->flat_indices = NULL;
    d->flat_count = 0;
    d->dirty = true;
    return w;
}